include/sys/queue.h

```diff
@@ -0,0 +1,842 @@
+#define __INCLUDE_SYS_QUEUE_H
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+#include <stdint.h>
+
+#ifdef CONFIG_ALLOW_BSD_COMPONENTS
+
+/* This file defines five types of data structures: singly-linked lists,
+ * lists, simple queues, tail queues and XOR simple queues.
+ *
+ *
+ * A singly-linked list is headed by a single forward pointer. The elements
+ * are singly linked for minimum space and pointer manipulation overhead at
+ * the expense of O(n) removal for arbitrary elements. New elements can be
+ * added to the list after an existing element or at the head of the list.
+ * Elements being removed from the head of the list should use the explicit
+ * macro for this purpose for optimum efficiency. A singly-linked list may
+ * only be traversed in the forward direction.  Singly-linked lists are ideal
+ * for applications with large datasets and few or no removals or for
+ * implementing a LIFO queue.
+ *
+ * A list is headed by a single forward pointer (or an array of forward
+ * pointers for a hash table header). The elements are doubly linked
+ * so that an arbitrary element can be removed without a need to
+ * traverse the list. New elements can be added to the list before
+ * or after an existing element or at the head of the list. A list
+ * may only be traversed in the forward direction.
+ *
+ * A simple queue is headed by a pair of pointers, one to the head of the
+ * list and the other to the tail of the list. The elements are singly
+ * linked to save space, so elements can only be removed from the
+ * head of the list. New elements can be added to the list before or after
+ * an existing element, at the head of the list, or at the end of the
+ * list. A simple queue may only be traversed in the forward direction.
+ *
+ * A tail queue is headed by a pair of pointers, one to the head of the
+ * list and the other to the tail of the list. The elements are doubly
+ * linked so that an arbitrary element can be removed without a need to
+ * traverse the list. New elements can be added to the list before or
+ * after an existing element, at the head of the list, or at the end of
+ * the list. A tail queue may be traversed in either direction.
+ *
+ * An XOR simple queue is used in the same way as a regular simple queue.
+ * The difference is that the head structure also includes a "cookie" that
+ * is XOR'd with the queue pointer (first, last or next) to generate the
+ * real pointer value.
+ *
+ * For details on the use of these macros, see the queue(3) manual page.
+ */
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define _Q_INVALID ((FAR void *)(uintptr_t)-1)
+#define _Q_INVALIDATE(a) (a) = _Q_INVALID
+
```

Comment:
In the original (I'm looking at the OpenBSD one here: [https://github.com/openbsd/src/blob/master/sys/sys/queue.h](https://github.com/openbsd/src/blob/master/sys/sys/queue.h)), `_Q_INVALID` is defined this way only when `defined(QUEUE_MACRO_DEBUG)` or `(defined(_KERNEL) && defined(DIAGNOSTIC)`. This makes the pointers -1 instead of NULL (0) when invalidating them for debugging purposes.

Here, we are defining `_Q_INVALID` and `_Q_INVALIDATE` unconditionally, which is not what I think we want.

Should this be wrapped with either a Kconfig or at least something like:

```
#if defined(CONFIG_DEBUG_SYS_QUEUE)
#  define _Q_INVALID ((void *)-1)
#  define _Q_INVALIDATE(a) (a) = _Q_INVALID
#else
#  define _Q_INVALIDATE(a)
#endif
```

(Though I generally prefer to reset unused pointers to NULL so that code can test whether they are invalid. So we could always define _Q_INVALIDATE(a) to _something_, perhaps `(a) = NULL` in the default case and `(a) = _Q_INVALID` in the debug case.)